pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java

```diff
@@ -0,0 +1,249 @@
+
+
+/**
+ * SingleValueAggregateRemoveRule that matches an Aggregate function SINGLE_VALUE and remove it
+ *
+ */
+public class PinotEvaluateLiteralRule extends RelOptRule {
+  public static final PinotEvaluateLiteralRule INSTANCE =
+      new PinotEvaluateLiteralRule(PinotRuleUtils.PINOT_REL_FACTORY);
+
+  public PinotEvaluateLiteralRule(RelBuilderFactory factory) {
+    super(operand(RelNode.class, any()), factory, null);
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    // Traverse the relational expression using a RexShuttle visitor
+    AtomicBoolean hasLiteralOnlyCall = new AtomicBoolean(false);
+    call.rel(0).accept(new RexShuttle() {
+      @Override
+      public RexNode visitCall(RexCall call) {
+        // Check if all operands are RexLiteral
+        if (call.operands.stream().allMatch(operand -> (operand instanceof RexLiteral))) {
+          // If all operands are literals, this call can be evaluated
+          hasLiteralOnlyCall.set(true);
+          // early terminate if we found one evaluate call
+          return call;
+        }
+        call.operands.forEach(operand -> {
+          if (hasLiteralOnlyCall.get()) {
+            // Early terminate if we already found one evaluate call
+            return;
+          }
+          // Recursively call visitCall on all RexCall operands
+          if (operand instanceof RexCall) {
+            visitCall((RexCall) operand);
+          }
+        });
+        // Return the call to continue traversal
+        return call;
+      }
+    });
+    return hasLiteralOnlyCall.get();
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    RelNode rootRelNode = call.rel(0);
+    RexBuilder rexBuilder = rootRelNode.getCluster().getRexBuilder();
+    Map<RexNode, RexNode> evaluatedResults = new HashMap<>();
+
+    // Recursively evaluate all the calls with Literal only operands from bottom up
+    // Traverse the relational expression using a RexShuttle visitor
+    rootRelNode.accept(new RexShuttle() {
+      @Override
+      public RexNode visitCall(RexCall call) {
+        call.operands.forEach(operand -> {
+          if (operand instanceof RexCall) {
+            visitCall((RexCall) operand);
+          }
+        });
+        // Check if all operands are RexLiteral
+        if (call.operands.stream().allMatch(operand -> operand instanceof RexLiteral)) {
+          // If all operands are literals or already evaluated, this call can be evaluated
+          return evaluateLiteralOnlyFunction(rexBuilder, call, evaluatedResults);
+        }
+        // Return the call to continue traversal
+        return call;
+      }
+    });
+
+    RelNode newRelNode = null;
+    if (rootRelNode instanceof Project) {
+      newRelNode = constructNewRelNode(rexBuilder, evaluatedResults, (Project) rootRelNode);
+    }
+    if (rootRelNode instanceof Filter) {
+      newRelNode = constructNewRelNode(evaluatedResults, (Filter) rootRelNode);
+    }
```

Comment:
Do we need to do this? I think the result of the visit is the new node
```suggestion
    RelNode newRelNode = rootRelNode.accept(new RexShuttle() {
      @Override
      public RexNode visitCall(RexCall call) {
        call.operands.forEach(operand -> {
          if (operand instanceof RexCall) {
            visitCall((RexCall) operand);
          }
        });
        // Check if all operands are RexLiteral
        if (call.operands.stream().allMatch(operand -> operand instanceof RexLiteral)) {
          // If all operands are literals or already evaluated, this call can be evaluated
          return evaluateLiteralOnlyFunction(rexBuilder, call, evaluatedResults);
        }
        // Return the call to continue traversal
        return call;
      }
    });
```